File: misc/shiren6dec/shiren6dec.cs
/*
Decrypts Shiren 6 Switch files (by bnnm)

compile: C:\Windows\Microsoft.NET\Framework\v4.0.30319\csc.exe /t:exe /out:shiren6dec.exe shiren6dec.cs
(needs +v4.0)

Files are encrypted using a manual implementation of AES-CTR using a fixed password + filename get salt.
This mostly adapts original C# code since it uses some arcane MS stuff like PasswordDeriveBytes.
Original decryptor is for a Filestream but this is simplified. There is probably some CTR mode from MS but who knows.

(I don't really know any C# so expect code to be wonky).
*/

using System;
using System.IO;
using System.Text;
using System.Security.Cryptography;

public class Decryptor : IDisposable {
	private static int BLOCK_SIZE = 0x10; //AES block

    private PasswordDeriveBytes pdb;
    private AesManaged aes;
    private ICryptoTransform cryptor;
    private FileStream src;
    private FileStream dst;
    //private CryptoStream dec;
	private byte[] blkBuffer = new byte[BLOCK_SIZE];
	private byte[] xorBuffer = new byte[BLOCK_SIZE];
	private UInt64 offset;
	private UInt64 blockNumber;

    public Decryptor(String filename, String name) {
        if (name == null) {
            name = Path.GetFileName(filename);
            name = Path.ChangeExtension(filename, null);
        }

        byte[] password = Encoding.UTF8.GetBytes("DtUp!43AZH46@*fj768GCM@ajNvEdBEB");
        byte[] salt = MakeSalt(name);

        // base AES-ECB and key used to create CTR xor pads
        pdb = new PasswordDeriveBytes(password, salt);

        aes = new AesManaged();
        aes.KeySize = 128; 
        aes.Mode = CipherMode.ECB; 
        aes.Padding = PaddingMode.None; 
        //aes.IV = (unused);
        aes.Key = pdb.GetBytes(aes.KeySize / 8); //each GetBytes returns a new key

        //cryptor = aes.CreateDecryptor(); //aes.Key, aes.IV
        cryptor = aes.CreateEncryptor(); //aes.Key, aes.IV
        
        src = new FileStream(filename, FileMode.Open);
        dst = new FileStream(filename + ".dec", FileMode.Create);
        //dec = new CryptoStream(src, decryptor, CryptoStreamMode.Read);
    }

    private byte[] MakeSalt(String name) {
        //while (name.Length < 8) {
        //    name += name;
        //}

        return Encoding.UTF8.GetBytes(name);
    }

	private void Reset() {
		offset = 0;
		blockNumber = (offset / (UInt64)BLOCK_SIZE) + 1;
	}

    public void Process() {
		byte[] input = new byte[BLOCK_SIZE];
		Reset();

        while (true) {
			int bytes = src.Read(input, 0, input.Length);	
			if (bytes <= 0)
				break;
			ProcessBlock(input, bytes);
			dst.Write(input, 0, bytes);

            /*
            int bytes = dec.Read(input, 0, input.Length);    
            if (bytes <= 0)
                break;
            dst.Write(input, 0, bytes);
            */
        }
    }
    
    // decrypt current block N (starts from 1)
    public void ProcessBlock(byte[] buf, int bytes) {
		int inputCount = blkBuffer.Length;
	 	//current num to LE array (example: int block 0x1122 = {0x22, 0x11, 0x00, 0x00}) then copy to full-sized array
		byte[] tmpBuffer = System.BitConverter.GetBytes(blockNumber);
		tmpBuffer.CopyTo(blkBuffer, 0);

		// encrypt current block number with filename-key + AES-ECB to get CTR xor.
		cryptor.TransformBlock(blkBuffer, 0, inputCount, xorBuffer, 0);
		blockNumber += 1;

		for (int i = 0; i < bytes; i++) {
			buf[i] ^= xorBuffer[i];
		}
	}

    public void Dispose() {
        if (pdb != null) pdb.Dispose();
        if (cryptor != null) cryptor.Dispose();
        if (aes != null) aes.Dispose();
        if (src != null) src.Dispose();
        if (dst != null) dst.Dispose();
        //if (dec != null) dec.Dispose();
    }

    public static void Main(String[] args) {
        String filename = null;
        String name = null;
        
        if (args == null || args.Length < 1) {
            Console.WriteLine("Shiren 6 decryptor by bnnm");
            Console.WriteLine("Usage: shiren6dec.exe filename");
            return;
        }

        if (args != null && args.Length > 0)
            filename = args[0];
        
        Decryptor dec = null;
        try {
            dec = new Decryptor(filename, name);

            Console.WriteLine("processing...");
            dec.Process();
            Console.WriteLine("done");
        } catch(System.IO.FileNotFoundException) {
            Console.WriteLine("File not found: " + filename);
        } finally {
            if (dec != null) dec.Dispose();
        }
        
    }
}

File: misc/shiren6dec/shiren6dec.py
# reimplementation of shiren6dec.cs b/c this is easier for me than .NET
# (although .NET version is easier to compare vs decompilations)

from Crypto.Cipher import AES
from Crypto.Util import Counter
from Crypto.Protocol import KDF
from Crypto.Hash import SHA1
import hashlib, glob, os

# semi-custom PBKDF1 (Rfc2898DeriveBytes is PBKDF2-HMAC-SHA-1)
#  KDF.PBKDF1(password, salt, length, count) seems to require 8-byte salt
#  https://github.com/microsoft/referencesource/blob/master/mscorlib/system/security/cryptography/passwordderivebytes.cs
#  https://stackoverflow.com/questions/27791726/encryption-diff-between-java-and-c-sharp/27798812#27798812
def MS_PasswordDeriveBytes(password, salt, key_size=16, iterations=100):
    # ComputeBaseValue
    hash = hashlib.sha1()
    hash.update(password)
    if salt:
        hash.update(salt)
    base_hash = hash.digest()

    for i in range(iterations - 1):
        base_hash = hashlib.sha1(base_hash).digest()
    out_hash = base_hash

    # ComputeBytes (not used when key_size=16)
    counter = 1
    while len(out_hash) < key_size:
        hash_counter = counter.to_bytes(4, byteorder='little')
        out_hash += hashlib.sha1(hash_counter + hash).digest()
        counter += 1
        print(counter, len(out_hash) < length)
    return out_hash[0: key_size]

# .NET AES CTR (as found in some games, not sure if standard)
# each block (starting from 1):
# - move counter to LE 0x10 array (block number 0x1122 = {0x22, 0x11, 0x00, 0x00})
# - encrypt array with AES-128-ECB
# - use resulting array to un-xor current 0x10 block
def decrypt_aes_ctr(data, key):
    ctr = Counter.new(128, little_endian=True, initial_value=1)
    cipher = AES.new(key, AES.MODE_CTR, counter=ctr)
    return cipher.decrypt(data)


password = b'DtUp!43AZH46@*fj768GCM@ajNvEdBEB'
for file in glob.glob('*.bundle'):
    name = file
    name = os.path.basename(name)
    name = os.path.splitext(name)[0]
    
    salt = name.encode('utf-8')
    key = MS_PasswordDeriveBytes(password, salt)
    

    with open(file, 'rb') as f:
        data = f.read()

    data = decrypt_aes_ctr(data, key)
    with open(file + '.dec', 'wb') as f:
        f.write(data)
